### DISCUSSION TITLE
How to add <script> snippet like Tag Manager to Nuxt 3 project?

### DISCUSSION THREAD

**Discussion by iannngit** [nuxt/nuxt#15968] [Questions] [answered]
I found in the doc [here](https://v3.nuxtjs.org/guide/features/head-management) introducing using <Head> <Script> but it doesn't seem working at all.

Anyone managed to place tag manager or analytics code snippets to Nuxt 3 project please?

```
(function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':
      new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],
    j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src=
    'https://www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f);
})(window,document,'script','dataLayer','GTM-ABCDEFGHIJKL');
```

**Comment by iannngit** [comment]


**Reply by atinux** [reply]
You can also do this in `app.vue` using

```ts
useHead({
  script: [
    { src:'https://www.googletagmanager.com/gtm.js?id=GTM-ABCDEFGHIJKL', async: true }
  ]
})
```

**Comment by iannngit** [comment, marked as answer]
OK I've found the correct solution to inject JS code snippets to <head> . 

```
export default defineNuxtConfig({
  css: [
      '@/assets/css/main.scss'
  ],
  scss: {
    "sourceMap": true
  },
  app: {
    head: {
      title: process.env.META_TITLE_GLOBAL,
      script: [
      {
        hid: 'gtm',
        children: `(function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':
      new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],
      j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src=
      'https://www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f);
      })(window,document,'script','dataLayer','${GOOGLE_TM_ID}');`,
        type: 'text/javascript'
      }
...
```

**Reply by iannngit** [reply]
This solution ensures GTM tags working. The way Atinux suggested is able to create a <script> in <head> however somehow tags don't work.

**Reply by viii47** [reply]
@iannngit that's because nuxt3 uses a different/newer module for the injection of meta and head data. This one expects a children property (as you've done) with the inner code, elements, ...script etc.

**Reply by andreasvirkus** [reply]
What about a `<noscript>` version of the tag for static sites, is there a way to add that as well to the start of the body tag?

If i add it to `app.vue`:

```html
<noscript>
  <iframe
    src="https://www.googletagmanager.com/ns.html?id=GTM-XXXXXX"
    height="0"
    width="0"
    style="display: none; visibility: hidden"
  ></iframe>
</noscript>
```

Then the iframe's tags get escaped:
```html
<noscript>
  &lt;iframe
    src=&quot;https://www.googletagmanager.com/ns.html?id=GTM-XXXXXX&quot;
    height=&quot;0&quot;
    width=&quot;0&quot;
    style=&quot;display: none; visibility: hidden&quot;
  &gt;&lt;/iframe&gt;
</noscript>
```

In a built/prod bundle:

![Screenshot 2022-08-26 at 14 43 54](https://user-images.githubusercontent.com/9140811/186896337-9a6497aa-7752-4c16-90f0-c4994f67c306.png)


Disabling JavaScript in the browser has no effect.

**Reply by andreasvirkus** [reply]
Seems this is an issue with `vueuse/head` package; opened an issue https://github.com/vueuse/head/issues/84

**Comment by windel07** [comment]
This works for me when I put `<Script>` tag inside `<Head>` in app.vue

<template v-if="config.public.gtmId">
    <Script
        :src="`https://www.googletagmanager.com/gtag/js?id=G-${config.public.gtmId}`"
        async
    />
    <Script
        :children="`window.dataLayer = window.dataLayer || []; function gtag(){dataLayer.push(arguments);} gtag('js', new Date());gtag('config', 'G-${config.public.gtmId}');`"
    />
</template>

**Comment by yesworld** [comment]
Hi there!

My **nuxt.config.ts,** how can I add the script to the end of the tag **</body>**? 
```ts
  script: [
                {
                    children: `(function(a,e,f,g,b,c,d){a[b]||(a.FintezaCoreObject=...........`,
                    type: 'text/javascript',
                    async: true,
                    mode: 'client',
                }
            ],
```

**Reply by sg208** [reply]
If you don't support older browsers, use "defer" instead of "async" even if the script is at the "head". It will fetch the code first but execute the script after everything else is loaded.

**Comment by Phyllon** [comment]
A dynamic example to this kind of question can be found here for Nuxt-3.0.0 (final): 
https://stackoverflow.com/a/75020196/19901666
https://stackoverflow.com/a/75006796/19901666

**Comment by toddpadwick** [comment]
I need to only inject the script if the cookies policy has been accepted – this was something the GTM module supported in Nuxt 2. Does anyone have a solution for injecting the script globally, programmatically when such an event occurs?
